File: pirellone/C++/skeleton.cpp
#include <cstdio>
#include <cstdlib>

int is_solvable(int M, int N, int **P);

void solve(int M, int N, int **P, void switch_row(int i), void switch_col(int j));

int main() {
    // checkpoint
    printf("%d\n", 0);
    // read choice
    static int choice;
    fflush(stdout);
    scanf("%d", &choice);
    // if choice {...} else {...}
    if (choice) {
        // read M, N
        static int M;
        static int N;
        fflush(stdout);
        scanf("%d%d", &M, &N);
        // for i to M {...}
        static int **P;
        P = new int*[M];
        for (int i = 0; i < M; i++) {
            // for j to N {...}
            P[i] = new int[N];
            for (int j = 0; j < N; j++) {
                // read P[i][j]
                fflush(stdout);
                scanf("%d", &P[i][j]);
            }
        }
        // call res = is_solvable(M, N, P)
        static int res;
        res = is_solvable(M, N, P);
        // write res
        printf("%d\n", res);
    } else {
        // read M, N
        static int M;
        static int N;
        fflush(stdout);
        scanf("%d%d", &M, &N);
        // for i to M {...}
        static int **P;
        P = new int*[M];
        for (int i = 0; i < M; i++) {
            // for j to N {...}
            P[i] = new int[N];
            for (int j = 0; j < N; j++) {
                // read P[i][j]
                fflush(stdout);
                scanf("%d", &P[i][j]);
            }
        }
        // call solve(M, N, P) callbacks {...}
        solve(M, N, P, [](int i) {
            // callback switch_row
            printf("%d %d\n", 1, 0);
            // write i
            printf("%d\n", i);
        }, [](int j) {
            // callback switch_col
            printf("%d %d\n", 1, 1);
            // write j
            printf("%d\n", j);
        });
        // no more callbacks
        printf("%d %d\n", 0, 0);
    }
    // exit
    exit(0);
}

File: pirellone/Rust/skeleton.rs
mod solution;
use std::io::Write;

macro_rules! readln {
    ($($var:expr),*) => {{
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        let parts: Vec<&str> = buf.trim().split(" ").collect();
        let mut i: usize = 0;
        $(
            assert!(i < parts.len(), "input format incorrect: too few values on this line"); 
            $var = parts[i].parse().unwrap();
            i += 1;
        )*
        assert!(i == parts.len(), "input format incorrect: too many values on the line")
    }};
}


fn main() {
    // checkpoint
    println!("{}", 0);
    // read choice
    let choice: i64;
    std::io::stdout().flush().unwrap();
    readln!(choice);
    // if choice {...} else {...}
    if choice != 0 {
        // read M, N
        let M: i64;
        let N: i64;
        std::io::stdout().flush().unwrap();
        readln!(M, N);
        // for i to M {...}
        let mut P: Vec<Vec<i64>> = Vec::new();
        P.resize(M as usize, vec::new());
        for i in 0..M {
            // for j to N {...}
            P[i as usize].resize(N as usize, 0);
            for j in 0..N {
                // read P[i][j]
                std::io::stdout().flush().unwrap();
                readln!(P[i as usize][j as usize]);
            }
        }
        // call res = is_solvable(M, N, P)
        let res: i64;
        res = solution::is_solvable(M, N, P);
        // write res
        println!("{}", res);
    } else {
        // read M, N
        let M: i64;
        let N: i64;
        std::io::stdout().flush().unwrap();
        readln!(M, N);
        // for i to M {...}
        let mut P: Vec<Vec<i64>> = Vec::new();
        P.resize(M as usize, vec::new());
        for i in 0..M {
            // for j to N {...}
            P[i as usize].resize(N as usize, 0);
            for j in 0..N {
                // read P[i][j]
                std::io::stdout().flush().unwrap();
                readln!(P[i as usize][j as usize]);
            }
        }
        // call solve(M, N, P) callbacks {...}
        solution::solve(M, N, P, |i: i64| {
            // callback switch_row
            println!("{} {}", 1, 0);
            // write i
            println!("{}", i);
        }, |j: i64| {
            // callback switch_col
            println!("{} {}", 1, 1);
            // write j
            println!("{}", j);
        });
        // no more callbacks
        println!("{} {}", 0, 0);
    }
    // exit
    std::process::exit(0);
}
